File: 2D.java
Time Complexity: O(log n)
Space Complexity: O(1)

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int m = matrix.length;
        int n = matrix[0].length;
        int low = 0;
        int high = m*n - 1;
        while(low <= high) {
            int mid = low + (high - low)/2;
            int row = mid/n;
            int col = mid%n;
            
            if(matrix[row][col] == target)
                return true;
            else if(matrix[row][col] > target)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return false;
    }
}

File: 2d.py
##Time Complexity : O(log(m)+log(n)) = O(log(mn))
##Space Complexity : O(1)
##Did this code successfully run on Leetcode : Yes

class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix:
        	return 0
        l, r = 0, len(matrix[0])-1

        while l < len(matrix) and r >= 0:
        	if matrix[l][r] == target:
        		return True 
        	elif matrix[l][r] < target:
        		l += 1
        	else:
        		r -= 1
        return False